driver/Src/CAN.c

```diff
@@ -0,0 +1,377 @@
+        */
+        init.Pin = GPIO_PIN_11;
+        init.Mode = GPIO_MODE_AF_PP;
+        init.Pull = GPIO_PULLUP;
+        init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        init.Alternate = GPIO_AF9_CAN1;
+        HAL_GPIO_Init(GPIOA, &init);
+
+        init.Pin = GPIO_PIN_12;
+        init.Mode = GPIO_MODE_AF_PP;
+        init.Pull = GPIO_NOPULL;
+        init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        init.Alternate = GPIO_AF9_CAN1;
+        HAL_GPIO_Init(GPIOA, &init);
+
+        // enable interrupts
+        HAL_NVIC_SetPriority(CAN1_TX_IRQn, 5, 0);
+        HAL_NVIC_EnableIRQ(CAN1_TX_IRQn);
+        HAL_NVIC_SetPriority(CAN1_RX0_IRQn, 5, 0);
+        HAL_NVIC_EnableIRQ(CAN1_RX0_IRQn);
+    }
+}
+
+// CAN MSP deinit
+void HAL_CAN_MspDeInit(CAN_HandleTypeDef *hcan) {
+    if (hcan->Instance == CAN1) {
+        // disable clocks
+        __HAL_RCC_CAN1_CLK_DISABLE();
+
+        /* disable gpio
+        PA11 -> CAN1_RX
+        PA12 -> CAN1_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_11);
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_12);
+
+        // disable interrupts
+        HAL_NVIC_DisableIRQ(CAN1_TX_IRQn);
+        HAL_NVIC_DisableIRQ(CAN1_RX0_IRQn);
+    }
+}
+
+can_status_t can_init(CAN_HandleTypeDef* handle, CAN_FilterTypeDef* filter) {
+    if (handle->Instance == CAN1) {
+        // init semaphores
+        can1_send_semaphore = xSemaphoreCreateBinaryStatic(
+            &can1_send_semaphore_buffer);
+        xSemaphoreGive(can1_send_semaphore);
+        can1_recv_semaphore = xSemaphoreCreateBinaryStatic(
+            &can1_recv_semaphore_buffer);
+        xSemaphoreGive(can1_recv_semaphore);
+
+        // init queues
+        can1_send_queue = xQueueCreateStatic(
+            CAN1_SEND_QUEUE_SIZE,
+            sizeof(tx_payload_t),
+            can1_send_queue_storage,
+            &can1_send_queue_buffer);
+        for (int i = 0; i < can1_recv_entry_count; i++) {
+            can1_recv_entries[i].queue = xQueueCreateStatic(
+                can1_recv_entries[i].size,
+                sizeof(rx_payload_t),
+                can1_recv_entries[i].storage,
+                &can1_recv_entries[i].buffer
+            );
+        }
+    } else {
+        return CAN_ERR;
+    }
+
+    // init HAL
+    if (HAL_CAN_Init(handle) != HAL_OK) {
+        return CAN_ERR;
+    }
+
+    // init filter
+    if (HAL_CAN_ConfigFilter(handle, filter) != HAL_OK) {
+        return CAN_ERR;
+    }
+
+    // enable interrupts
+    if (HAL_CAN_ActivateNotification(handle, CAN_IT_TX_MAILBOX_EMPTY) != HAL_OK) {
+        return CAN_ERR;
+    }
+    if (HAL_CAN_ActivateNotification(handle, CAN_IT_RX_FIFO0_MSG_PENDING) != HAL_OK)
+    {
+        return CAN_ERR;
+    }
+
+    // start CAN
+    if (HAL_CAN_Start(handle) != HAL_OK) {
+        return CAN_ERR;
+    }
+
+    initialized = true;
+    return CAN_OK;
+}
+
+can_status_t can_recv(CAN_HandleTypeDef* handle, uint16_t id, CAN_RxHeaderTypeDef* header, uint8_t data[], bool blocking) {
```

Comment:
Why not just return the full rx_payload_t struct to the user rather than header and data separately?

Reply:
I would like it to match the HAL.

Also, I like the idea of someone having a struct that describes their data:

```
struct brake_light_msg {
    bool on;
}

struct brake_light;
```

and then them using that as their data

```
can_recv(handle, id, header, (uint8_t[])brake_light, blocking);
```

Reply:
i suppose that makes sense